policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+        assert not self.service_watcher
+        assert not self.pod_watcher
+        assert not self.endpoints_watcher
+
+        self.namespace_watcher = Thread(target=_keep_watch,
+                                        args=(self.watcher_queue, "namespaces", self.nsResourceVersion))
+        self.service_watcher = Thread(target=_keep_watch,
+                                      args=(self.watcher_queue, "services", self.svcResourceVersion))
+        self.pod_watcher = Thread(target=_keep_watch,
+                                  args=(self.watcher_queue, "pods", self.poResourceVersion))
+        self.endpoints_watcher = Thread(target=_keep_watch,
+                                        args=(self.watcher_queue, "endpoints", self.epResourceVersion))
+
+        self.namespace_watcher.daemon = True
+        self.service_watcher.daemon = True
+        self.pod_watcher.daemon = True
+        self.endpoints_watcher.daemon = True
+
+    def read_updates(self):
+        """
+        Pulls an update off the Queue and processes it.
+        If no responses remain in the queue, it will trigger a resync for all resource lists.
+        """
+        try:
+            response = self.watcher_queue.get_nowait()
+            r_json = json.loads(response)
+            r_type = r_json["type"]
+            r_obj = r_json["object"]
+            r_kind = r_obj["kind"]
+
+            _log.debug("Reading update %s: %s" % (r_type, r_kind))
+
+            if r_kind in VALID_KINDS:
+                if r_type == CMD_DELETED:
+                    self.delete_resource(kind=r_kind, target=r_obj)
+                elif r_type in [CMD_ADDED, CMD_MODIFIED]:
+                    self.process_resource(action=r_type,
+                                          kind=r_kind,
+                                          target=r_obj)
+                else:
+                    _log.error("Event from watch not recognized: %s" % r_type)
+            else:
+                _log.error("Resource %s not Pod, Service, Endpoints or Namespace" % r_kind)
+        except Queue.Empty:
+            self.resync()
+
+    def process_resource(self, action, kind, target):
+        """
+        Takes a target object and an action and updates internal resource pools
+        :param action: String of [CMD_ADDED, CMD_MODIFIED] (returned by api watch)
+        :param kind: String in VALID_KINDS
+        :param target: json dict of resource info
+        """
+        # Determine Resource Pool
+        if kind == KIND_NAMESPACE:
+            resource_pool = self.changed_namespaces
+            obj = Namespace(target)
+        elif kind == KIND_SERVICE:
+            resource_pool = self.changed_services
+            obj = Service(target)
+        elif kind == KIND_POD:
+            resource_pool = self.changed_pods
+            obj = Pod(target)
+        elif kind == KIND_ENDPOINTS:
+            resource_pool = self.changed_endpoints
+            obj = Endpoints(target)
+        else:
```

Comment:
Rather than have an else branch here, let's just assert that kind is in VALID_KINDS at the beginning of this method.  If we pass an invalid kind here, that is a bug in our code, so we should fail loudly.